### ISSUE TITLE
Should `destroyObject` also replace `.destroy()`

### ISSUE THREAD

**Issue by jjspace** [CesiumGS/cesium#12266]
While reviewing https://github.com/CesiumGS/cesium/pull/12129 I finally noticed the `if !isDestoryed() > .destroy()` pattern and it felt very odd to me.
```js
if (!this._environmentMapManager.isDestroyed()) {
  this._environmentMapManager.destroy();
}
```

My [comment](https://github.com/CesiumGS/cesium/pull/12129#discussion_r1797134135):

> It feels like the responsibility for this check should live in the `destroy` function itself. I feel like you should be able to just call `destroy()` regardless and let it handle checking if it's already destroyed and essentially turn into a noop function. Then in this function the logic would just be "We know we want to guarantee this is destroyed so call `destroy()`" and no wrapper check for if it's already destroyed?

Looking at the [`destroyObject`](https://github.com/CesiumGS/cesium/blob/192ab0687518feba009b5055537b5ab8b3afc523/packages/engine/Source/Core/destroyObject.js#L35) function I see we delete all the properties and replace the `.isDestroyed` function with a function that only returns true.

https://github.com/CesiumGS/cesium/blob/192ab0687518feba009b5055537b5ab8b3afc523/packages/engine/Source/Core/destroyObject.js#L53

I think a potentially easy path forward to streamline this pattern and fit the way I _expected_ it to work would be to also (or maybe instead of) replace the `.destroy()` function with a noop `() => {}`.

This would allow us to rewrite that original example as just one line and not have to care about checking first. The responsibility is with the original object even though it's all been deleted by `destoryObject`
```js
this._environmentMapManager.destroy();
```

**Comment by jjspace** [comment]
> `// If the property is destroyable, then call "destroyObject" on this property as well:`
>     `if (key === "destroy") {`
If this is the only alternative we can come up with then I don't think it should happen. As you pointed out this opens up a lot of unknowns about properties that may or may not exist. Also the potential to call destroy on something that's in a property but not considered fully "owned" by the object being destroyed.

> The `destroy` function is also supposed to destroy all destroyable properties as well:

Replacing the `destroy` function in `destroyObject` would not prevent this functionality from happening. Calling `SomeClass.destroy()` the first time would call the proper function which handles destroying all of the objects it owns and needs to destroy. Then it would call `destroyObject` which would replace the `destroy` function with a no-op meaning the next time you call `SomeClass.destroy()` it does nothing. But there should also be nothing it _has_ to do anymore since it already tore down everything it should have.

**Comment by javagl** [comment]
> If this is the only alternative we can come up with then I don't think it should happen. 

Yes, the side-effects are hard to predict here, and that approach would require a concept of "ownership" that is _much_ more strict than what is even possible in JS. 

> Replacing the destroy function in destroyObject would not prevent this functionality from happening. 

I think I misunderstood one detail there. 
But to summarize: The answer to the question in the title, from my perspective, is a clear "Yes!"

Generally speaking, calling "destroy" on a destroyed object should be safe and a no-op. In other languages, this is solved with some 
```java
final void destroy() {
    if (isDestroyed()) return;
    reallyDestroy(); // To be implemented by subclasses
}
```
But given that it's possible to juggle around `function` objects here (and this is already happening for all other functions anyhow), replacing `destroy` with `() => {}` would make sense IMHO.

**Comment by donmccurdy** [comment]
I've run into a snag with #13452, converting `Texture.js` to an ES6 class. This loop...

https://github.com/CesiumGS/cesium/blob/192ab0687518feba009b5055537b5ab8b3afc523/packages/engine/Source/Core/destroyObject.js#L47-L51

... no longer iterates over methods of ES6 classes, because `Object.keys(texture)` does not return them. We can instead do something like...

```javascript
for (const key of Object.getOwnPropertyNames(Object.getPrototypeOf(object))) {
  if (typeof object[key] === "function") { 
    object[key] = throwOnDestroyed; 
  } 
}
```

... but that won't include _inherited_ methods, if the class extends a parent class. But I suspect the previous method didn't include inherited methods, either. We could 'fix' `destroyObject(o)` to handle ES6 classes, and optionally inherited methods by traversing up the prototype chain. But I am wondering if that's what we want? Modifying instances this way seems like a bit of a [performance risk](https://github.com/CesiumGS/cesium/issues/3754#issuecomment-4192362052), though I'm not sure how we'd confirm that's a real-world problem...

I don't have any particular opinion on whether calling .destroy() twice should be an error or a silent no-op, but replacing all the methods on an ES6 class instance feels like it might be overkill, particularly if calling .destroy() twice is allowed.

**Comment by javagl** [comment]
I do _not_ have strong opinions about _how certain things are implemented_, because there might be aspects that I don't fully have on the radar. But a few thoughts in terms of the usage:

Right now, there is this pattern of 
`primitive = primitive && !primitive.isDestroyed() && primitive.destroy();`
which is convoluted. Being able to simplify that to
`primitive &&= primitive.destroy();`
could be a small win. Consequently, I think that calling `destroy()` _twice_ should not be an error. 

Beyond that, the questions about "ownership" that have been discussed above make this subtle, and they are not modeled cleanly and explicitly enough. On a green field, one could think about an `interface Destructible` (or whether it might be called `Destroyable`) that defines the `destroy()` and `isDestroyed()` methods, and explains the semantics. One could brainstorm whether this interface should also offer a `Destroyable[] getOwnedObjects()` so that the management (aka destruction) of these could happen from the "outside", or (what I'd prefer:) to leave that to the instances itself, so that it can have some `private ownedObjects: Destroyable[]` that is filled in the constructor and purged in the `destroy()` implementation, or whether it handles these objects manually and explicitly.

**Comment by donmccurdy** [comment]
Based on @jjspace's comments above it, sounds like he also would prefer that calling `destroy()` twice should not be an error — I'm comfortable with that direction, then!

IMO, it follows from that decision that if `destroy()` can be called after a class is destroyed, other methods on the class do not need to be replaced with error-throwing stubs either? If the existing methods do happen to throw errors because they depend on now-deleted properties, that's fine. If there are specific cases where the lack of errors would be disruptive or confusing, we could use this pattern...

```javascript
class MyClass {
  someImportantMethod() {
    //>>includeStart('debug', pragmas.debug);
    assert(!this.isDestroyed(), 'Object destroyed');
    //>>includeEnd('debug');

    ...
  }
}
```

... but we wouldn't want to do that in every method, everywhere.

It's not much used yet, but there is this:

https://github.com/CesiumGS/cesium/blob/c42bbc073f7eda8e950385506aa18e11bc3f5b59/packages/engine/Source/Core/globalTypes.js#L11-L13

In retrospect I probably should have used `@interface` rather than `@typedef`, but that's an easy change.

> One could brainstorm whether ... the management (aka destruction) of these could happen from the "outside", or (what I'd prefer:) to leave that to the instances itself...

I'd prefer the same. Resource ownership is too complex to centralize in a context-less external function, I expect, and better handled by the class itself.

**Comment by donmccurdy** [comment]
Proposed changes. Let's leave non-ES6 classes as-is. As we gradually move to ES6 classes, let's:

1. Remove `destroyObject` calls from ES6 classes
2. Enforce that the class tracks `._destroyed` state and implements `.destroy()` / `.isDestroyed()` by itself
3. Enforce that `.destroy()` is [idempotent](https://en.wikipedia.org/wiki/Idempotence)
4. Enforce 1–3 automatically with ast-grep and/or unit tests, and/or update `destroyObject` to warn if given an ES6 class instance as input

**Comment by donmccurdy** [comment]
So as not to block the adoption of ES6 classes and type checking, I've opened a PR extending `destroyObject` with support for ES6 class instances as input:

- #13560

Per comments above, I'd still be in favor of some changes to `destroyObject`, but maybe that's best kept separate from ES6 class adoption.
